lib/rucio/common/types/dicts/replica.py

```diff
@@ -0,0 +1,44 @@
+# -*- coding: utf-8 -*-
+# Copyright European Organization for Nuclear Research (CERN) since 2012
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, List
+from .typed_dict import TypedDict
+
+
+class BadReplicaDict(TypedDict):
+    scope: str
+    name: str
+    rse_id: str
+
+
+class ReplicaPFNDict(TypedDict):
+    rse: str
+    rse_id: str
+    type: str
+    domain: str
+    priority: int
+    volatile: bool
+    client_extract: bool
+
+
+class ReplicaDict(TypedDict):
+    scope: str
+    name: str
+    bytes: int
+    md5: str
+    adler32: str
+    rses: Dict[str, List[str]]
+    pfns: Dict[str, Dict[str, ReplicaPFNDict]]
+    states: Dict[str, str]
```

Comment:
I'm not sure about naming for these classes. Function parameters that use `BadReplicaDict` are usually called `pfns`, but is there a more general term that includes both PFN and a RSE-specific replica? `Replica` is also a somewhat unclear term whether it refers to a specific file on a single RSE or the replication of a file across multiple RSEs, any suggestions for differentiating the names? 

The goal of these classes is to re-use them across different functions. A problem I can imagine appearing is if functions take slightly different versions of the same dict (maybe some keys are optional), which would require duplicating the dicts for the different functions. I'm not sure if it will be a problem though.

Reply:
See `ReplicaClient.declare_bad_file_replicas` for instance.

Reply:
It's becoming quite the overlap between some slightly-different dicts now. Some potential solutions:
- Use subclassing where possible, for instance every dict with `scope: str` and `name: str` could inherit from `DidDict`
- Define TypedDict privately just before the corresponding function definitions if they are only used once. Not ideal since using TypedDicts often mean annotating variable assignments explicitly (otherwise just `Dict` is assumed), which means that they have to be public. Could still be done in-line with the function definitions, but it is also a bit messy. 
- Ignore it. It is not a new problem with having this many different dictionaries anyway, they have always been there, just not this explicit.

Reply:
I am leaning to just ignore it for now. The root of evil here is that we give the same name to actual different things. So this would have to be "fixed" in the first place, but I am not even sure if it is fix-worthy. In the scope of type checking I would ignore it and just have different dicts.